clients/src/main/java/org/apache/kafka/common/serialization/ListSerializer.java

```diff
@@ -0,0 +1,98 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.serialization;
+
+import org.apache.kafka.clients.CommonClientConfigs;
+import org.apache.kafka.common.config.ConfigException;
+import org.apache.kafka.common.utils.Utils;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+public class ListSerializer<Inner> implements Serializer<List<Inner>> {
+
+    private Serializer<Inner> inner;
+    private boolean isFixedLength;
+
+    static private List<Class<? extends Serializer>> fixedLengthSerializers = Arrays.asList(
+        ShortSerializer.class,
+        IntegerSerializer.class,
+        FloatSerializer.class,
+        LongSerializer.class,
+        DoubleSerializer.class,
+        UUIDSerializer.class);
+
+    public ListSerializer() {}
+
+    public ListSerializer(Serializer<Inner> serializer) {
+        this.inner = serializer;
+        this.isFixedLength = fixedLengthSerializers.contains(serializer.getClass());
+    }
+
+    @SuppressWarnings(value = "unchecked")
+    @Override
+    public void configure(Map<String, ?> configs, boolean isKey) {
+        if (inner == null) {
+            final String innerSerdePropertyName = isKey ? CommonClientConfigs.DEFAULT_LIST_KEY_SERDE_INNER_CLASS : CommonClientConfigs.DEFAULT_LIST_VALUE_SERDE_INNER_CLASS;
+            final Object innerSerde = configs.get(innerSerdePropertyName);
+            try {
+                if (innerSerde instanceof String) {
+                    inner = Utils.newInstance((String) innerSerde, Serde.class).serializer();
+                } else if (innerSerde instanceof Class) {
+                    inner = ((Serde<Inner>) Utils.newInstance((Class) innerSerde)).serializer();
+                } else {
+                    throw new ClassNotFoundException();
+                }
+                inner.configure(configs, isKey);
+            } catch (final ClassNotFoundException e) {
+                throw new ConfigException(innerSerdePropertyName, innerSerde, "Serde class " + innerSerde + " could not be found.");
+            }
+        }
+    }
+
+    @Override
+    public byte[] serialize(String topic, List<Inner> data) {
+        if (data == null) {
+            return null;
+        }
+        final int size = data.size();
+        try (final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+             final DataOutputStream out = new DataOutputStream(baos)) {
+            out.writeInt(size);
+            for (Inner entry : data) {
+                final byte[] bytes = inner.serialize(topic, entry);
```

Comment:
inner.serialize() can return null here in case of the list entries is null.
then NPE will follow at https://github.com/apache/kafka/pull/6592/files#diff-9333dde189fe982b61d2f01aa3b2cba8R83

Reply:
Hmmm that's an interesting edge case.
I cannot just return null since a list might contain real values i.e. `List<String> data = {'A', null, 'C'}`
I have to serialize `null` somehow...

Reply:
I get it with AbstractKafkaAvroSerializer and my custom Array serde

AbstractKafkaAvroSerializer code: 
```
    protected byte[] serializeImpl(String subject, Object object) throws SerializationException {
        Schema schema = null;
        if (object == null) {
            return null;
        } else {
            String restClientErrorMsg = "";
```

Reply:
That is a tricky question. There are multiple ways how we encode this, but this seem to be a design question that required to go back to the KIP discussion.

For example, we could skip the optimization of fixed-length types and encode the length for every entry -- a length of `-1` would indicate a `null`. Or we introduce a "header" that tells us if there are `null` in the list (either a bit-array for short list or a "list of null positions")

Example for list of null positions would be: `<numberOfNulls><listOfNullPositions><regularEncoding>` 
Ie, with the example for above, we encode `1-1-<bytes-for-A>-<bytes-for-C>`.

As bit array, it would be `<bitArryaSizeInBytes><bitArray><regularEncoding>`
Ie, with the example for above, we encode `1-0100000-<bytes-for-A>-<bytes-for-C>`

Reply:
@yeralin I was thinking about this case more. I really think, that the fix-length optimization is valuable as it reduced the serialized byte size by 50% for e.g. Integer lists. For the other two proposals, it's harder to judge which one is better. I see the following (dis)advantages for each:

null-index-list:
 - low overhead for dense lists with few nulls (for zero nulls, it’s 4 byte overhead, for each null, it’s additional 4 bytes)
 - the longer the lists, the smaller the overhead

bit-array:
 - low overhead for short lists (4 bytes for byte array length + list-lenght/8 bytes for the byte-array itself)
 - not ideal for long lists with few nulls

Hence, for short list both might be equally ok. For long lists, it depends of they are dense or sparse (for long-dense list, null-index-list seems to be better, for long-sparse-lists, bit-array seems to be better).

It will be hard to tell which one is better, hence, I would suggest that we only implement the null-index-list list for now, because I assume that dense lists are more common and it works better for long dense lists than the bit-array idea.

However, to allow us to support different serialization format in the future, we should add one more magic byte in the very beginning that encodes the choose serialization format. In our case, we will will have one format and the magic byte will always be "zero". If we add the byte-array format, we can just set the magic byte to one to indicate the other format. And we could even add more formats of people have a better idea how to do it later on.

Btw: we could actually already go with two formats:
 - 0 => optimized-fixed-length-encoding plus null-index-list
 - 1 => variable length encoding using `-1` in the length field to indicate `null` (no header to mark nulls is required at all for this case).

If we agree on this design, we should update the KIP accordingly.

\cc @vvcephei: would love to hear your feedback, too.

Reply:
Thank you for your input @mjsax 

Your outline makes sense, and I'm happy to jump back to KIP.

However, could we please wrap up this round of review? Make sure that all generics, docs, tests are in place.

Once it is clean and polished, I can start updating KIP and implementing this new feature. 

What do you think?